apps/openssl.c

```diff
@@ -76,6 +77,8 @@ static void calculate_columns(DISPLAY_COLUMNS *dc)
 
 static int apps_startup(void)
 {
+    setlocale(LC_ALL, "");
```

Comment:
Problem with this of course is that you're making suggestion that all library applications should do that. What if application doesn't, would libcrypto still work? And how?

Reply:
Well, the key question was actually "how" would it work. And not in sense "good" vs. "bad", but whether or not it would be producing varying results depending on whether or not application called setlocale(LC_ALL,""). Should it be concern, or would we be bettor of it behaviour was invariant?

Reply:
If the input string to `mbstowcs` doesn't agree with the locale, `mbstowcs` will return `(size_t)-1`, so using *that* variant will simply be invalidated...  however, it currently bails out when this happens, that's probably not the right move...

Reply:
Just note that calling setlocale() in libcrypto would definitely NOT be the right thing to do as it is not thread safe and some applications might want to set different locale. It is probably obvious but I just wanted to explicitly state that.

Reply:
And yes, if mbstowcs fails, the code should just skip the variant and not bail out with error.

Reply:
Re setlocale(): I suspected that.  Thanks for confirming.

> And yes, if mbstowcs fails, the code should just skip the variant and not bail out with error.

Yup, I've fixed that

Reply:
To summarize. libcrypto shall not call setlocale. Remaining question if it should make assumption that application did. [I for one would answer "no, it should not."]

Reply:
> Remaining question if it should make assumption that application did. [I for one would answer "no, it should not."]

Ok, so

1. How do you detect that?  Simply assume that if the current locale is "C", the application didn't set locale?
2. What should be done if we determine the application didn't set it?

Reply:
(BTW, I don't quite agree with you, I think it *is* reasonable to assume that locale is correctly set, and to fail variant 2 if it isn't, i.e. if mbstowcs fails.

Reply:
> I think it is reasonable to assume that locale isexpect the locale to be correctly set

To "" or "some-other-string"? Here is fun Linux-specific question. What if non-UTF-8 locales for CLI application and xterm are different? Or one is UTF-8 and another is not...

Reply:
> I cannot judge what an application uses for a locale, as long as text strings are encoded to match.

Exactly my point. And in my book it translates to "I don't want to be dependent on application to call setlocale, be it with "" or "some-non-empty-string"." All you can do is to assume that it's all sorted out by application and/or user [choosing adequate input method], and act in manner that is not dependent on what application code did [or didn't]. To be more specific don't call mbstowcs :-)